compiler/modules/front/codeGen.bal

```diff
@@ -1311,52 +1314,89 @@ function codeGenTypeTest(CodeGenContext cx, bir:BasicBlock bb, Environment env,
     return { result, block: nextBlock, narrowing };   
 }
 
-function codeGenCheckingExpr(CodeGenContext cx, bir:BasicBlock bb, Environment env, s:CheckingKeyword checkingKeyword, s:Expr expr) returns CodeGenError|ExprEffect {
+function codeGenCheckingStmt(CodeGenContext cx, bir:BasicBlock bb, Environment env, s:CheckingKeyword checkingKeyword, s:Expr expr) returns CodeGenError|StmtEffect {
+    // checking stmt falls into one of : 1) never err 2) always err 3) conditionally err
     var { result: o, block: nextBlock } = check codeGenExpr(cx, bb, env, expr);
     // Constants should be resolved during constant folding
-    bir:Register operand = <bir:Register>o; 
+    bir:Register operand = <bir:Register>o;
     t:SemType errorType =  t:intersect(operand.semType, t:ERROR);
+    bir:BasicBlock block;
+    t:SemType resultType;
     if t:isNever(errorType) {
-        return { result: operand, block: nextBlock };
+        block = nextBlock;
+        resultType = operand.semType;
     }
     else {
-        bir:Register isError = cx.createRegister(t:BOOLEAN);
-        bir:TypeTestInsn typeTest = { operand, semType: t:ERROR, result: isError, negated: false };
-        nextBlock.insns.push(typeTest);
-        bir:InsnRef testInsnRef = bir:lastInsnRef(nextBlock);
-        bir:BasicBlock okBlock = cx.createBasicBlock();
-        bir:BasicBlock errorBlock = cx.createBasicBlock();
-        bir:CondBranchInsn condBranch = { operand: isError, ifTrue: errorBlock.label, ifFalse: okBlock.label };
-        nextBlock.insns.push(condBranch);
-        bir:Register errorReg = cx.createRegister(errorType);
-        bir:CondNarrowInsn narrowToError = {
-            result: errorReg,
-            operand,
-            basis: { insn: testInsnRef, result: true }
-        };
-        errorBlock.insns.push(narrowToError);
-        if checkingKeyword == "check" {
-            bir:RetInsn insn = { operand: errorReg };
-            errorBlock.insns.push(insn);
-        }
-        else {
-            bir:PanicInsn insn = { operand: errorReg };
-            errorBlock.insns.push(insn);
+        resultType = t:diff(operand.semType, t:ERROR);
+        if t:isNever(resultType) {
+            codeGenCheckingTerminal(nextBlock, checkingKeyword, operand);
+            return { block: () };
         }
-        t:SemType okType = t:diff(operand.semType, t:ERROR);
-        if t:isNever(okType) {
-            // this has to be an error because otherwise the register would be of type never
-            // SUBSET fix this if we allow checking keyword as a call statement
+        // JBUG build fails when the order of binding pattern is reversed
+        { block, result: _ } = check codeGenCheckingCond(cx, nextBlock, operand, errorType, checkingKeyword, resultType);
+    }
+    // resultType === NEVER case is already handled
+    if resultType !== t:NIL {
+        return cx.semanticErr(`type of ${checkingKeyword} statement must be nil or never`);
```

Comment:
It's better to phrase this in terms of the type of the operand i.e. operand of ${checkingKeyword} must be a subtype of error?